%\VignetteIndexEntry{GGtools overview}
%\VignetteDepends{}
%\VignetteKeywords{Genetical genomics,SNP,expression}
%\VignettePackage{GGtools}

%
% NOTE -- ONLY EDIT THE .Rnw FILE!!!  The .tex file is
% likely to be overwritten.
%
\documentclass[12pt]{article}

\usepackage{amsmath,pstricks}
\usepackage[authoryear,round]{natbib}
\usepackage{hyperref}


\textwidth=6.2in
\textheight=8.5in
%\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}


\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}

\textwidth=6.2in

\bibliographystyle{plainnat} 
 
\usepackage{/Users/stvjc/WOrk/ExternalSoft/R-devel/share/texmf/Sweave}
\begin{document}
%\setkeys{Gin}{width=0.55\textwidth}

\title{Overview of \Rpackage{GGtools} for genetical genomics}
\author{VJ Carey \texttt{<stvjc@channing.harvard.edu>}}
\maketitle

\section{Introduction}

We use the term \textit{genetical genomics} to refer to
data analysis activities that link genotypic information
such as SNP configurations to gene expression phenotype.
The \Rpackage{GGtools} package includes various
demonstration resources and analysis tools for these
activities.
We will attach the library and have a look at a basic
demonstration resource.
\begin{Schunk}
\begin{Sinput}
> library(GGtools)
> data(c20GGceu)
> class(c20GGceu)
\end{Sinput}
\begin{Soutput}
[1] "ggExprSet"
attr(,"package")
[1] ".GlobalEnv"
\end{Soutput}
\begin{Sinput}
> c20GGceu
\end{Sinput}
\begin{Soutput}
GG Expression Set (exprSet catering for many SNP attributes) with 
	8793 genes
	48 samples
There are  114666  attributes; names include:
rs4814683 rs6076506 rs6139074 rs1418258 rs7274499 
\end{Soutput}
\end{Schunk}
The ggExprSet class is an extension of the
\Rclass{exprSet} class.  It represents expression
data from the hgfocus chip on 48 individuals in the CEU
CEPH group, and SNP data obtained from their HapMap genotyping
results.

The data are organized into an 8793
by 48 matrix of expression values
accessible with the \texttt{exprs} method, and
an 48 by 114666
of rare allele counts:
\begin{Schunk}
\begin{Sinput}
> dim(exprs(c20GGceu))
\end{Sinput}
\begin{Soutput}
[1] 8793   48
\end{Soutput}
\begin{Sinput}
> dim(snps(c20GGceu))
\end{Sinput}
\begin{Soutput}
[1]     48 114666
\end{Soutput}
\begin{Sinput}
> snps(c20GGceu)[1:5, 1:5]
\end{Sinput}
\begin{Soutput}
        rs4814683 rs6076506 rs6139074 rs1418258 rs7274499
NA11829         0         2         0         0         2
NA11830         1         2         1         1         1
NA11831         0         2         0         0         2
NA11832         1         2         1         1         2
NA11839         0         2         0         0         2
\end{Soutput}
\end{Schunk}

We need some genetic metadata about SNPs; these are
culled from SNP genotyping panels released on a 
chromosome-by-chromosome basis for CEPH participants
by HapMap project:
\begin{Schunk}
\begin{Sinput}
> data(chr20meta)
> chr20meta[1:4, ]
\end{Sinput}
\begin{Soutput}
            pos strand
rs4814683  9795      +
rs6076506 11231      +
rs6139074 11244      +
rs1418258 11799      +
\end{Soutput}
\end{Schunk}

A basic task is to compute a screen (over the genome, or,
more practically, over a chromosome) of genotypic determination
of expression.  The \Rfunction{detScreen} function help
with this; defaults are supplied for an example related to
results in Cheung and Spielman 2005:

\begin{center}
\begin{Schunk}
\begin{Sinput}
> tmp = as(chr20meta, "data.frame")
> chr20meta = data.frame(rsnum = tmp, tmp)
> ds1 = detScreen()