%\VignetteIndexEntry{GGtools overview}
%\VignetteDepends{}
%\VignetteKeywords{Genetical genomics,SNP,expression}
%\VignettePackage{GGtools}

%
% NOTE -- ONLY EDIT THE .Rnw FILE!!!  The .tex file is
% likely to be overwritten.
%
\documentclass[12pt]{article}

\usepackage{amsmath,pstricks}
\usepackage[authoryear,round]{natbib}
\usepackage{hyperref}


\textwidth=6.2in
\textheight=8.5in
%\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}


\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}

\textwidth=6.2in

\bibliographystyle{plainnat} 
 
\usepackage{/Users/stvjc/WOrk/ExternalSoft/R-devel/share/texmf/Sweave}
\begin{document}
%\setkeys{Gin}{width=0.55\textwidth}

\title{Overview of \Rpackage{GGtools} for genetical genomics}
\author{VJ Carey \texttt{<stvjc@channing.harvard.edu>}}
\maketitle

\section{Introduction}

We use the term \textit{genetical genomics} to refer to
data analysis activities that link genotypic information
such as SNP configurations to gene expression phenotype.
The \Rpackage{GGtools} package includes various
demonstration resources and analysis tools for these
activities.
We will attach the library and have a look at a basic
demonstration resource.
\begin{Schunk}
\begin{Sinput}
> library(GGtools)
\end{Sinput}
\begin{Soutput}
Loading required package: Biobase
Loading required package: tools
Loading required package: hgfocus
Loading required package: GGdata
Loading required package: geneplotter
Loading required package: annotate
KernSmooth 2.22 installed
Copyright M. P. Wand 1997
Loading required package: RColorBrewer
\end{Soutput}
\begin{Sinput}
> data(chr20GGceu)
> class(chr20GGceu)
\end{Sinput}
\begin{Soutput}
[1] "racExSet"
attr(,"package")
[1] "GGtools"
\end{Soutput}
\begin{Sinput}
> chr20GGceu
\end{Sinput}
\begin{Soutput}
racExSet instance (SNP rare allele count + expression)
rare allele count assayData:
  Storage mode: environment 
  featureNames: rs4814683, rs6076506, rs6139074, ..., rs6062370, rs6090120 (117417 total)
  Dimensions:
          racs
Rows    117417
Samples     58

expression assayData
  Storage mode: environment 
  featureNames: 1007_s_at, 1053_at, 117_at, ..., AFFX-r2-P1-cre-3_at, AFFX-r2-P1-cre-5_at (8793 total)
  Dimensions:
        exprs
Rows     8793
Samples    58

phenoData
  sampleNames: NA06985, NA06993, NA06994, ..., NA12891, NA12892 (58 total)
  varLabels and descriptions:
    sample: arbitrary numbering

Experiment data
  Experimenter name: Cheung VG 
  Laboratory: Department of Pediatrics, University of Pennsylvania, Philadelphia, Pennsylvania 19104, USA. vcheung@mail.med.upenn.edu 
  Contact information:  
  Title: Mapping determinants of human gene expression by regional and genome-wide association. 
  URL:  
  PMIDs: 16251966 

  Abstract: A 180 word abstract is available. Use 'abstract' method.

Annotation [1] "hgfocus"
\end{Soutput}
\end{Schunk}
The racExSet class is an extension of the
\Rclass{eSet} class.  It represents expression
data from the hgfocus chip on 48 individuals in the CEU
CEPH group, and SNP data obtained from their HapMap genotyping
results.

The data are organized into an 8793
by 58 matrix of expression values
accessible with the \texttt{exprs} method, and
an 117417 by 58
of rare allele counts:
\begin{Schunk}
\begin{Sinput}
> dim(exprs(chr20GGceu))
\end{Sinput}
\begin{Soutput}
[1] 8793   58
\end{Soutput}
\begin{Sinput}
> dim(snps(chr20GGceu))
\end{Sinput}
\begin{Soutput}
[1] 117417     58
\end{Soutput}
\begin{Sinput}
> snps(chr20GGceu)[1:5, 1:5]
\end{Sinput}
\begin{Soutput}
          NA06985 NA06993 NA06994 NA07000 NA07022
rs4814683       2       0       0       2       1
rs6076506       0       0       0       0      NA
rs6139074       2       0       0       2       1
rs1418258       2       0       0       2       1
rs7274499       0       0       0       0      NA
\end{Soutput}
\end{Schunk}

We need some genetic metadata about SNPs; these are
culled from SNP genotyping panels released on a 
chromosome-by-chromosome basis for CEPH participants
by HapMap project:
\begin{Schunk}
\begin{Sinput}
> data(chr20meta)
> chr20meta[1:4, ]
\end{Sinput}
\begin{Soutput}
            pos strand
rs4814683  9795      +
rs6076506 11231      +
rs6139074 11244      +
rs1418258 11799      +
\end{Soutput}
\end{Schunk}

A basic task is to compute a screen (over the genome, or,
more practically, over a chromosome) of genotypic determination
of expression.  The \Rfunction{snpScreen} method helps
with this; we illustrate an example related to
results in Cheung and Spielman 2005:

\begin{Schunk}
\begin{Sinput}
> S100 = snpScreen(chr20GGceu, chr20meta, genesym("CPNE1"), ~., 
+     lm, gran = 25)
> S100
\end{Sinput}
\begin{Soutput}
GGtools snpScreenResult for call:
snpScreen(racExSet = chr20GGceu, snpMeta = chr20meta, gene = genesym("CPNE1"), 
    formTemplate = ~., fitter = lm, gran = 25)
There were 4697 attempted fits,
and 4697 were successful.
\end{Soutput}
\end{Schunk}
A primitive display is obtained as follows.  We know that \texttt{lm}
was used, so the relevant p-values are in the coefficient component
of the summarized fit objects.
\begin{Schunk}
\begin{Sinput}
> ps = as.numeric(sapply(S100, function(x) try(summary(x)$coef[2, 
+     4])))
> plot(S100@locs, -log10(ps), xlab = "chromosomal location")
\end{Sinput}
\end{Schunk}
\includegraphics{GGoverview-dofig}

%\begin{center}
%<<dosc,fig=TRUE,results=hide>>=
%ss1 = snpScreen(chr20GGceu, 

%The \Rfunction{detScreen} function returns a list of fit results:
%<<lkds>>=
%names(ds1)
%ds1[[1]]
%length(ds1[[2]])
%names(ds1[[2]])
%ds1[[2]][[1]][[1]]

\section{Performance-oriented specialization}

The \texttt{snpScreen} method illustrated above is very
general (can accommodate and retain results of any R modeling function)
but fairly slow.  We have added an R function \texttt{fastAGM}
for fast fitting of an additive genetic model (equivalent to
but much faster than using \texttt{lm}).  We apply this
for IRF5 on chromosome 7, and attempt to fit on all SNPs that
have no missing values:
\begin{Schunk}
\begin{Sinput}
> ut = unix.time(sIRF5 <- snpScreen(chr7GGceu, chr7meta, genesym("IRF5"), 
+     ~., fastAGM, 1))
> ut
\end{Sinput}
\begin{Soutput}
[1] 47.161  6.239 81.018  0.000  0.000
\end{Soutput}
\begin{Sinput}
> sIRF5
\end{Sinput}
\begin{Soutput}
GGtools snpScreenResult for call:
snpScreen(racExSet = chr7GGceu, snpMeta = chr7meta, gene = genesym("IRF5"), 
    formTemplate = ~., fitter = fastAGM, gran = 1)
There were 128625 attempted fits,
and 86328 were successful.
\end{Soutput}
\begin{Sinput}
> which.min(extract_p(sIRF5))
\end{Sinput}
\begin{Soutput}
rs6969930 
   104180 
\end{Soutput}
\end{Schunk}
\begin{Schunk}
\begin{Sinput}
> plot_mlp(sIRF5, chr7meta)
\end{Sinput}
\end{Schunk}
 

\section{Appendix: Package documentation for \Rpackage{GGtools}}
\begin{Schunk}
\begin{Soutput}
		Information on package 'GGtools'

Description:


Package:       GGtools
Title:         software and data for genetical genomics (c) 2006 VJ
               Carey
Version:       0.3-0
Author:        stvjc <stvjc@channing.harvard.edu>
Description:   dealing with hapmap SNP reports, GWAS, etc.
Depends:       R (>= 2.2.0), Biobase (>= 1.11.26), hgfocus, GGdata,
               geneplotter, RColorBrewer
Maintainer:    stvjc <stvjc@channing.harvard.edu>
License:       Artistic (see COPYING)
Packaged:      Thu Aug 10 11:02:04 2006; stvjc
Built:         R 2.4.0; powerpc-apple-darwin8.7.0; 2006-08-16 01:11:21;
               unix


Index:


HM2rac                  compute rare allele count from a hapmap file
detScreen               perform a screen for SNP determination of
                        expression using a ggExprSet instance, and
                        produce a plot
geneLocs                gene metadata from NCBI
genoStrings             create a character vector of genotype value
                        strings
ggExprSet-class         Class "ggExprSet", representation of genetical
                        genomics (expression + SNP) data
ggreg                   regression analysis for additive effect of
                        genotype on expression
ggrplot                 plot a regression model for additive effect of
                        genotype on expression
make_racExSet           create a racExSet from simpler constituents
racExSet-class          Class "racExSet" for combining RareAlleleCount
                        representations of SNPs, gene expression data,
                        and other phenotype data
regseq                  compute a sequence of regression models
                        relating genotype to expression
snpScreen               compute model fits over a sequence of SNPs
snps                    accessor for genotype data in a ggExprSet


Further information is available in the following vignettes in
directory '/Users/stvjc/WOrk/ExternalSoft/R-devel/library/GGtools/doc':


GGoverview: GGtools overview (source)
\end{Soutput}
\end{Schunk}

Session information for this vignette build:
\begin{Schunk}
\begin{Sinput}
> sessionInfo()
\end{Sinput}
\begin{Soutput}
R version 2.4.0 Under development (unstable) (2006-07-16 r38625) 
powerpc-apple-darwin8.7.0 

locale:
C

attached base packages:
[1] "tools"     "methods"   "stats"     "graphics"  "grDevices" "utils"    
[7] "datasets"  "base"     

other attached packages:
     GGtools RColorBrewer  geneplotter     annotate       GGdata      hgfocus 
     "0.3-0"      "0.2-3"     "1.11.4"     "1.11.2"      "0.0-2"     "1.12.0" 
     Biobase 
   "1.11.26" 
\end{Soutput}
\end{Schunk}

\end{document}
